internal/controller/reinstall/reinstall.go

```diff
@@ -372,3 +379,279 @@ func (r *ReinstallHandler) finalizeReinstallRequest(
+
+	err = preservation.Backup(ctx, r.Client, log, clusterInstance)
+	if err != nil {
+		log.Error("Encountered error during data preservation", zap.Error(err))
+
+		preservationDataBackedupCondition = reinstallPreservationDataBackedupConditionStatus(
+			metav1.ConditionFalse, v1alpha1.Failed, err.Error())
+
+		clusterIdentityDetectedCondition = reinstallClusterIdentityDataDetectedConditionStatus(
+			metav1.ConditionFalse, v1alpha1.Failed, "Data preservation failed")
+
+		return
+	}
+
+	log.Info("Data preservation completed, computing backup summary...")
+
+	// Add a small delay to allow for the last resource preserved to be created before computing stats
+	time.Sleep(1 * time.Second)
+
+	preservationDataBackedupCondition, clusterIdentityDetectedCondition, err =
+		getDataPreservationSummary(ctx, r.Client, log, clusterInstance)
+
+	return
+}
+
+// ensurePreservedDataIsRestored restores preserved resources
+func (r *ReinstallHandler) ensurePreservedDataIsRestored(
+	ctx context.Context,
+	log *zap.Logger,
+	clusterInstance *v1alpha1.ClusterInstance,
+) (result reconcile.Result, err error) {
+
+	result = reconcile.Result{}
+	err = nil
+
+	var preservationDataRestoredCondition *metav1.Condition
+
+	defer func() {
+		if preservationDataRestoredCondition != nil {
+			patch := client.MergeFrom(clusterInstance.DeepCopy())
+
+			if changed := setReinstallStatusCondition(clusterInstance, *preservationDataRestoredCondition); changed {
+				result.Requeue = true
+				updateErr := conditions.PatchCIStatus(ctx, r.Client, clusterInstance, patch)
+				if updateErr != nil {
+					result.RequeueAfter = requeueWithShortInterval
+					err = logAndWrapUpdateFailure(log, err, updateErr,
+						fmt.Sprintf("failed to update reinstall restore data condition '%s'",
+							preservationDataRestoredCondition.Type))
+				}
+			}
+		}
+	}()
+
+	if clusterInstance.Spec.Reinstall.PreservationMode == v1alpha1.PreservationModeNone {
+
+		preservationDataRestoredCondition = reinstallPreservationDataRestoredConditionStatus(
+			metav1.ConditionFalse, v1alpha1.PreservationNotRequired, "PreservationMode is set to None.")
+
+		return
+	}
+
+	preservationDataRestoredCondition = findReinstallStatusCondition(clusterInstance,
+		v1alpha1.ReinstallPreservationDataRestored)
+
+	if preservationDataRestoredCondition == nil ||
+		preservationDataRestoredCondition.Reason == string(v1alpha1.Initialized) {
+		preservationDataRestoredCondition = reinstallPreservationDataRestoredConditionStatus(
+			metav1.ConditionFalse, v1alpha1.InProgress, "Data restoration in progress")
+		return
+	}
+
+	switch preservationDataRestoredCondition.Reason {
+	case string(v1alpha1.Failed):
+		err = errors.New(preservationDataRestoredCondition.Message)
+		return
+
+	case string(v1alpha1.Completed), string(v1alpha1.PreservationNotRequired), string(v1alpha1.DataUnavailable):
+		// Nothing to do:
+		// restore was completed successfully, or was not required, or no preservation data was found
+		return
+	}
```

Comment:
Same here why check the conditions? Just make sure any preserved data is restored and it should be a no-op if you're done already.

Reply:
If we decide to change the backup logic - we can apply the same here.

Reply:
Refactored to run preservation & restore each time.